packages/firestore/scripts/prune-dts.ts

```diff
@@ -0,0 +1,420 @@
+          type.expression
+        );
+        if (
+          publicSymbol &&
+          publicSymbol.name !== currentName &&
+          verifyMatchingTypeArguments(
+            type,
+            publicSymbol.declarations[0] as
+              | ts.ClassDeclaration
+              | ts.InterfaceDeclaration
+          )
+        ) {
+          // If there is a public type that we can refer to, update the import
+          // statement to refer to the public type.
+          exportedTypes.push(
+            ts.updateExpressionWithTypeArguments(
+              type,
+              type.typeArguments,
+              ts.createIdentifier(publicSymbol.name)
+            )
+          );
+        } else {
+          // Hide the type we are inheriting from and merge its declarations
+          // into the current class.
+          const privateType = typeChecker.getTypeAtLocation(type);
+          additionalMembers.push(
+            ...convertPropertiesForEnclosingClass(
+              program,
+              host,
+              sourceFile,
+              privateType.getProperties(),
+              node
+            )
+          );
+        }
+      }
+    }
+
+    if (exportedTypes.length > 0) {
+      prunedHeritageClauses.push(
+        ts.updateHeritageClause(heritageClause, exportedTypes)
+      );
+    }
+  }
+
+  if (ts.isClassDeclaration(node)) {
+    return ts.updateClassDeclaration(
+      node,
+      node.decorators,
+      node.modifiers,
+      node.name,
+      node.typeParameters,
+      prunedHeritageClauses,
+      [...node.members, ...(additionalMembers as ts.ClassElement[])]
+    ) as T;
+  } else if (ts.isInterfaceDeclaration(node)) {
+    return ts.updateInterfaceDeclaration(
+      node,
+      node.decorators,
+      node.modifiers,
+      node.name,
+      node.typeParameters,
+      prunedHeritageClauses,
+      [...node.members, ...(additionalMembers as ts.TypeElement[])]
+    ) as T;
+  } else {
+    throw new Error('Only classes or interfaces are supported');
+  }
+}
+
+/**
+ * Iterates the provided symbols and returns named declarations for these
+ * symbols if they are missing from `currentClass`. This allows us to merge
+ * class hierarchies for classes whose inherited types are not part of the
+ * public API.
+ *
+ * This method relies on a private API in TypeScript's `codefix` package.
+ */
+function convertPropertiesForEnclosingClass(
+  program: ts.Program,
+  host: ts.CompilerHost,
+  sourceFile: ts.SourceFile,
+  parentClassSymbols: ts.Symbol[],
+  currentClass: ts.ClassDeclaration | ts.InterfaceDeclaration
+): ts.NamedDeclaration[] {
+  const newMembers: ts.NamedDeclaration[] = [];
+  // The `codefix` package is not public but it does exactly what we want. We
+  // can explore adding a slimmed down version of this package to our repository
+  // if this dependency should ever break.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  (ts as any).codefix.createMissingMemberNodes(
```

Comment:
Does it handle generics with different names? For example:
```
declare class A<T> {
  aData: T;
}

export declare class B<K> extends A<K> {
  bData:K;
}
```
Does it become:
```
export declare class B<K> {
  aData: K; // instead of aData: T
  bData: K;
}
```